datasource/ethereum/src/block_stream.rs

```diff
@@ -753,49 +841,66 @@ where
-                // If found in store
-                if let Some(local_block) = local_block_opt {
-                    Box::new(future::ok(local_block))
-                } else {
-                    let ctx_clone1 = ctx;
-                    let ctx_clone2 = ctx_clone1.clone();
-
-                    // Request from Ethereum node instead
-                    Box::new(
-                        ctx_clone1
-                            .eth_adapter
-                            .block_by_hash(&ctx_clone1.logger, block_hash)
+        let eth = self.eth_adapter.clone();
+        let logger = self.logger.clone();
+        
+        // Search for the block in the store first then use the ethereum adapter as a backup
+        let block = future::result(ctx.chain_store.block(block_hash))
+            .and_then(move |local_block_opt| -> Box<Future<Item = _, Error = _> + Send> {
+                match local_block_opt {
+                    Some(block) => Box::new(future::ok(block)),
+                    None => {
+                        let logger = ctx.logger.clone();
+                        let ctx_1 = ctx.clone();
+                        let ctx_2 = ctx_1.clone();
+                        Box::new(ctx_1.eth_adapter
+                            .block_by_hash(&logger, block_hash)
                             .and_then(move |block_opt| {
                                 block_opt.ok_or_else(move || {
-                                    format_err!(
-                                        "Ethereum node could not find block with hash {}",
-                                        block_hash
-                                    )
+                                    format_err!("Ethereum node could not find block with hash {}", block_hash)
                                 })
                             })
                             .and_then(move |block| {
-                                ctx_clone1
+                                ctx_1
                                     .eth_adapter
-                                    .load_full_block(&ctx_clone1.logger, block)
+                                    .load_full_block(&logger, block)
                                     .map_err(|e| format_err!("Error loading full block: {}", e))
                             })
                             .and_then(move |block| {
                                 // Cache in store for later
-                                ctx_clone2
+                                ctx_2
                                     .chain_store
                                     .upsert_blocks(stream::once(Ok(block.clone())))
                                     .map(move |()| block)
-                            }),
-                    )
+                            }))
+                    }
                 }
-            },
-        )
+            })
+            .and_then(move |block| -> Box<Future<Item = _, Error = _> + Send>{
+                if !include_calls_in_block {
```

Comment:
This `include calls -> either return the block without calls or fetch the calls` logic is showing up in a few places. Could be worth consolidating into a single helper function/method.